ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m

```diff
@@ -111,43 +111,87 @@ +(ADAuthenticationResult*) resultFromCancellation: (NSUUID*) correlationId
+                                                              errorDetails: @"No broker response received."];
+    return [ADAuthenticationResult resultFromError:error correlationId:nil];
+}
+
++ (ADAuthenticationResult*)resultForBrokerErrorResponse:(NSDictionary*)response
+{
+    NSString* correlationIdStr = [response valueForKey:OAUTH2_CORRELATION_ID_RESPONSE];
+    NSUUID* correlationId =  correlationIdStr ? [[NSUUID alloc] initWithUUIDString:correlationIdStr] : nil;
     
-    NSUUID* correlationId = [response valueForKey:OAUTH2_CORRELATION_ID_RESPONSE] ?
-                            [[NSUUID alloc] initWithUUIDString:[response valueForKey:OAUTH2_CORRELATION_ID_RESPONSE]]
-                            : nil;
+    // Otherwise parse out the error condition
+    ADAuthenticationError* error = nil;
     
-    if(!response || [response valueForKey:OAUTH2_ERROR_DESCRIPTION])
+    NSString* errorDetails = [response valueForKey:OAUTH2_ERROR_DESCRIPTION];
+    if (!errorDetails)
     {
-        ADAuthenticationError* error = nil;
-        NSString* errorDetails = nil;
-        NSInteger errorCode = 0;
-        if (response)
-        {
-            errorDetails = [response valueForKey:OAUTH2_ERROR_DESCRIPTION];
-            errorCode = [[response valueForKey:@"error_code"] integerValue];
-            
-            if (!errorDetails)
-            {
-                errorDetails = @"Broker did not provide any details";
-            }
-        }
-        else
-        {
-            errorDetails = @"No broker response received.";
-        }
-        
-        error = [ADAuthenticationError errorFromNSError:[NSError errorWithDomain:ADBrokerResponseErrorDomain code:errorCode userInfo:nil] errorDetails:errorDetails];
+        errorDetails = @"Broker did not provide any details";
+    }
         
-        return [ADAuthenticationResult resultFromError:error correlationId:correlationId];
+    NSString* strErrorCode = [response valueForKey:@"error_code"];
+    NSInteger errorCode = AD_ERROR_BROKER_UNKNOWN;
+    if (strErrorCode && ![strErrorCode isEqualToString:@"0"])
+    {
+        errorCode = [strErrorCode integerValue];
+    }
+    
+    NSString* protocolCode = [response valueForKey:@"protocol_code"];
+    if (!protocolCode)
+    {
+        // Older brokers used to send the protocol code as "code" and the error code not at all
+        protocolCode = [response valueForKey:@"code"];
+    }
+    
+    if (![NSString adIsStringNilOrBlank:protocolCode])
+    {
+       
+        error = [ADAuthenticationError errorFromAuthenticationError:errorCode
+                                                       protocolCode:protocolCode
+                                                       errorDetails:errorDetails];
+    }
+    else
+    {
+        NSError* nsError = [NSError errorWithDomain:ADBrokerResponseErrorDomain
+                                               code:errorCode
+                                           userInfo:nil];
```

Comment:
This is actually the indent style you're supposed to use on Objective-C, the colons ( : ) line up, unless you're seeing something I don't.